### DISCUSSION TITLE
Importing Bitwarden Database

### DISCUSSION THREAD

**Discussion by Friday13th87** [keepassxreboot/keepassxc#6228] [Q&A]
Is there any good way to import a Bitwarden (csv) Database in current KeepassXC Version?
I tried importing my database several times but keepassxc is always freezing while importing.
Tried it in debian as flatpak and appimage installation (current version) and as apt installation (current debian buster stabel version) and under windows 10 with current version with a powerful pc (Ryzen 2700x, 32GB RAM, nvme)
And is there a way to preserve my Bitwarden Folders/Groups as well as TOTP entries?

Thanks in advance

**Comment by nathan-denyer** [comment]
I was able to import successfully from Bitwarden using the CSV importer tool and selecting the correct headers (including TOTP/folders). KeepassXC also freezes when importing a Bitwarden CSV for me on Windows. I exited the app after the import when it froze, and upon reopening everything appears to be imported correctly. I'm not sure if there's an issue importing the Bitwarden specific CSV or any CSV file.

**Comment by droidmonkey** [comment]
We only support CSV import at this time.

**Comment by Friday13th87** [comment]
> 
> 
> We only support CSV import at this time.

i know, but i dont understand the sense of this post

**Comment by droidmonkey** [comment]
Sorry I read your post as a feature request instead of a usage question.

**Comment by DinizBortolotto** [comment]
Today I installed KeepassXC and tried to import Bitwarden Exported CSV.
But after reordering columns and click Done, it seems that KeepassXC is disregarding columns chosen/reordered.
I will try again later and keep track of steps to reproduce and maybe open a bug report.

**Comment by droidmonkey** [comment]
Can you provide a sample Bitwarden CSV file for our testing purposes?

**Comment by DinizBortolotto** [comment]
> Today I installed KeepassXC and tried to import Bitwarden Exported CSV.
> But after reordering columns and click Done, it seems that KeepassXC is disregarding columns chosen/reordered.
> I will try again later and keep track of steps to reproduce and maybe open a bug report.

0 - Removed KeePassXC config and cache, pretending to be a fresh install. 
Bitwarden's CSV is already downloaded.
1 - Started KeePassXC at console and confirmed (say yes to) "check for updates on startup".
2 - Click on "Import from CSV" and selected CSV file.
3 - Fill "Database Name" and "Description" and clicked "Continue".
4 - Keep Encryption Settings unchanged and clicked "Continue".
5 - Entered a Password at "Database Credentials", clicked "Done" and saved Passwords.kbdx at $HOME.

6 - KeePassXC now shows "Import CSV field" dialog with 10 (ten) columns from Bitwarden's CSV that not fit KeePassXC's order or associations. (PrtSc attached)
![keepassxc_importCSV](https://user-images.githubusercontent.com/4468840/110085637-19f18580-7d70-11eb-87dd-d753733611bb.png)

7 - I did some adjusts like reordering and removed some columns. (PrtSc - before check "First line has field names" combo - attached)
![keepassxc_importCSV_fixed](https://user-images.githubusercontent.com/4468840/110085727-38578100-7d70-11eb-854a-13e783321077.png)

8 - And TA-dah, gotcha! (PrtSc attached)
![keepassxc_importCSV_after_remove_header](https://user-images.githubusercontent.com/4468840/110087073-f0395e00-7d71-11eb-8e4e-bbc8317ec5fc.png)

When we click on "First line has field names" combo-box the adjustments made at step 7 are all gone.
I didn't notice that last time I tried to import CSV and clicked Done. 

I don't know if this is expected behavior, but I think if user has already changed columns order/selection it shouldn't be changed/undone. Should I open a new bug for this or this is a normal behavior?

**Comment by ikesaber** [comment]
Hello @droidmonkey et al, 

I wanted to check in and ask if there's been any capabilities change for migrating from Bitwarden to keepassxc. Is the process still CSV-only? have there been any improvements?

**Reply by droidmonkey** [reply]
It is still csv only. Do they have another export that is more complete?

**Reply by ikesaber** [reply]
@droidmonkey sorry for the delayed response - i guess i don't get notifications when somebody replies to me.

I'm not sure what you mean by "another export that is more complete" their export will typically look like this (when opened as a spreadsheet):

| Cloud | favorite | type | name | notes | fields | reprompt | login_uri | login_username | login_password | login_totp |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| label i picked | (can be blank) | login | google.com | i like google | First Name: john </br> </br> Last Name: doe </br> </br> (you can add as many of these as you want - they appear as separate fields similar to username or password) | 0 or 1 | https://www.google.com | john.doe@gmail.com | password1 | (i don't use this but i assume it looks like any other otp cred.) |

It does look like it has more configurable options (fields, for example - i don't know if there's an analog for this in keepassxc)

**Reply by michaelk83** [reply]
> Do they have another export that is more complete?

https://github.com/keepassxreboot/keepassxc/issues/8367#issuecomment-1214590103
